File: leetcode/april2024/23_minimum_height_trees.cpp
// CODE

class Solution {
    public:
    vector<int> findMinHeightTrees(int n, vector<vector<int>>& edges) {
        if(n == 1) return {0};
        vector<vector<int>> adjList(n);
        vector<int> indegree(n, 0), res;

        // creating adjacency matrix
        for(auto &e : edges) {
            adjList[e[0]].push_back(e[1]);
            adjList[e[1]].push_back(e[0]);
            indegree[e[0]]++;
            indegree[e[1]]++;
        }

        // adding leaf nodes into queue
        queue<int> q;
        for(int i = 0; i < n; i++)
            if(indegree[i] == 1) q.push(i), indegree[i]--;
        
        while(!q.empty()) {
            int s = q.size();
            res.clear();
            for(int i = 0; i < s; i++) {
                int curr = q.front(); q.pop();
                res.push_back(curr);
                for(auto nbr : adjList[curr]) {
                    indegree[nbr]--;
                    if(indegree[nbr] == 1) q.push(nbr);
                }
            }
        }

        return res;
    }
};

File: leetcode/april2024/python/23_minimum_height_trees.py
# CODE

class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1:
            return [0]
        
        adj_list = [[] for _ in range(n)]
        indegree = [0] * n
        res = []

        # creating adjacency list and calculating indegree
        for e in edges:
            adj_list[e[0]].append(e[1])
            adj_list[e[1]].append(e[0])
            indegree[e[0]] += 1
            indegree[e[1]] += 1

        # adding leaf nodes into queue
        q = deque()
        for i in range(n):
            if indegree[i] == 1:
                q.append(i)
                indegree[i] -= 1

        while q:
            s = len(q)
            res.clear()
            for _ in range(s):
                curr = q.popleft()
                res.append(curr)
                for nbr in adj_list[curr]:
                    indegree[nbr] -= 1
                    if indegree[nbr] == 1:
                        q.append(nbr)
        
        return res